src/test/net_tests.cpp

```diff
@@ -395,6 +395,60 @@ BOOST_AUTO_TEST_CASE(cnetaddr_basic)
     BOOST_CHECK(!addr.SetSpecial("totally bogus"));
 }
 
+BOOST_AUTO_TEST_CASE(cnetaddr_tostring_canonical_ipv6)
+{
+    // Test that CNetAddr::ToString formats IPv6 addresses with zero compression as described in
+    // RFC 5952 ("A Recommendation for IPv6 Address Text Representation").
+    const std::map<std::string, std::string> canonical_representations_ipv6{
+        {"0000:0000:0000:0000:0000:0000:0000:0000", "::"},
+        {"000:0000:000:00:0:00:000:0000", "::"},
+        {"000:000:000:000:000:000:000:000", "::"},
+        {"00:00:00:00:00:00:00:00", "::"},
+        {"0:0:0:0:0:0:0:0", "::"},
+        {"0:0:0:0:0:0:0:1", "::1"},
+        {"2001:0:0:1:0:0:0:1", "2001:0:0:1::1"},
+        {"2001:0db8:0:0:1:0:0:1", "2001:db8::1:0:0:1"},
```

Comment:
`2001:0db8:0:0:1:0:0:1` can be represented as `2001:db8::1:0:0:1`, but also as `2001:0db8:0:0:1::1`.

Are we guaranteed that all [`getnameinfo()`](https://linux.die.net/man/3/getnameinfo) on different platforms will use the variants in this test? I think no.

(similarly in some others below)

Reply:
RFC 5952 tells us that there is one (and only one) canonical representation of `2001:0db8:0:0:1:0:0:1` which is `2001:db8::1:0:0:1`.

Testing conformance with RFC 5952 is the entire point of this PR :)

Reply:
> RFC 5952 tells us that there is one (and only one) canonical representation

Where is that text?

I see the following, which seems to be the opposite:

https://tools.ietf.org/html/rfc5952#section-2.2

> 'The "::" can only appear once in an address.'
   This gives a choice on where in a single address to compress the
   zero.
      2001:db8::aaaa:0:0:1
      2001:db8:0:0:aaaa::1

Reply:
@vasild 

The text you're quoting is describing the problem.

The RFC describes how to solve that problem using canonicalisation :)

See the first paragraph of the RFC: "This document defines a canonical textual representation format."

Note that "canonical textual representation" in this context means one (and only one) way to format it.

Perhaps reading https://en.wikipedia.org/wiki/IPv6_address#Representation makes the role of canonicalisation of IPv6 addresses more clear? :)

Reply:
My bad, https://tools.ietf.org/html/rfc5952#section-4.2.3 is pretty clear. Sorry for the noise.